addons/gamification/models/gamification_goal.py

```diff
@@ -0,0 +1,403 @@
+    last_update = fields.Date(
+        help="In case of manual goal, reminders are sent if the goal as not been updated for a while (defined in challenge). Ignored in case of non-manual goal or goal not linked to a challenge.")
+
+    definition_description = fields.Text(related='definition_id.description', readonly=True)
+    definition_condition = fields.Selection(related='definition_id.condition', readonly=True)
+    definition_suffix = fields.Char(related='definition_id.full_suffix', string="Suffix", readonly=True)
+    definition_display = fields.Selection(related='definition_id.display_mode', string="Display Mode", readonly=True)
+
+    @api.depends('definition_condition', 'current', 'target_goal')
+    def _compute_completeness(self):
+        """Return the percentage of completeness of the goal, between 0 and 100"""
+        # res = dict.fromkeys(ids, 0.0)
+        for goal in self:
+            if goal.definition_condition == 'higher':
+                if goal.current >= goal.target_goal:
+                        goal.completeness = 100.0
+                else:
+                    goal.completeness = round(100.0 * goal.current / goal.target_goal, 2)
+            elif goal.current < goal.target_goal:
+                # a goal 'lower than' has only two values possible: 0 or 100%
+                goal.completeness = 100.0
+            else:
+                goal.completeness = 0.0
+
+    @api.onchange('definition_id')
+    def _onchange_definition_id(self):
+        goal_definition = self.definition_id
+        self.computation_mode = goal_definition.computation_mode
+        self.definition_condition = goal_definition.condition
+
+    @api.model
+    def create(self, vals):
+        """Overwrite the create method to add a 'no_remind_goal' field to True"""
+        return super(GamificationGoal, self.with_context(no_remind_goal=True)).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        """Overwrite the write method to update the last_update field to today
+
+        If the current value is changed and the report frequency is set to On
+        change, a report is generated
+        """
+        vals['last_update'] = fields.Date.today()
+        result = super(GamificationGoal, self).write(vals)
+        for goal in self:
+            if goal.state != "draft" and ('definition_id' in vals or 'user_id' in vals):
+                # avoid drag&drop in kanban view
+                raise UserError(_('Can not modify the configuration of a started goal'))
+
+            if vals.get('current'):
+                if 'no_remind_goal' in self.env.context:
+                    # new goals should not be reported
+                    continue
+
+                if goal.challenge_id and goal.challenge_id.report_message_frequency == 'onchange':
+                    goal.challenge_id.sudo().report_progress(users=goal.user_id.ids)
```

Comment:
This will cause to traceback, report_progress's users param now considered as recordset, either send goal.user_id or change report_progress to deal with this(I would prefer to send recordset), also check the docstring of report_progress method